_source/_posts/2022-03-30-java-regex.md

```diff
@@ -0,0 +1,254 @@
+
+Regular expressions make finding patterns in text much easier. Some high-level use cases include:
+* Email validation
+* Phone number validation
+* Credit card number validation
+* Password pattern matching
+* String replacement
+
+Regular expressions are also well-supported in many programming languages.
+
+## How to use regular expressions in Java
+
+Supporting classes for regular expressions in Java are available in the `java.util.regex` package. As per [Java's documentation](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html), the package primarily consists of the three classes below.
+
+**`Pattern`:** The object of this class is a compiled representation of a regular expression. There are no public constructors available in this class, so you need to call the `compile()` static method to create an object.
+
+Here are the different implementation details and method signature of the `compile()` method available in the `Pattern` class as per Java's documentation.
+
+ ```java
+   /**
+* Compiles the given regular expression into a pattern.
+*
+* @param  regex
+*         The expression to be compiled
+* @return the given regular expression compiled into a pattern
+* @throws  PatternSyntaxException
+*          If the expression's syntax is invalid
+*/
+public static Pattern compile(String regex) {
+    return new Pattern(regex, 0);
+}
+/**
+* Compiles the given regular expression into a pattern with the given
+* flags.
+* @param  regex
+*         The expression to be compiled
+*
+* @param  flags
+*         Match flags, a bit mask that may include
+*         {@link #CASE_INSENSITIVE}, {@link #MULTILINE}, {@link #DOTALL},
+*         {@link #UNICODE_CASE}, {@link #CANON_EQ}, {@link #UNIX_LINES},
+*         {@link #LITERAL}, {@link #UNICODE_CHARACTER_CLASS}
+*         and {@link #COMMENTS}
+*
+* @return the given regular expression compiled into a pattern with the given flags
+*/
+public static Pattern compile(String regex, int flags) {
+    return new Pattern(regex, flags);
+}
+ ```
+
+**`Matcher`:** This class also has no public constructor, and the object of this class is constructed by invoking the `matcher()` method on a `Pattern` class object. The `Matcher` object will match the given input against a given pattern.
+
+**`PatternSyntaxException`:** This class throws an unchecked exception to indicate a syntax error in a regular-expression pattern.
+
+The typical code flow of working with a regular expression in Java is shown below.
+
+We first create a `Pattern` object by invoking its static `compile()` method, and then we pass it a pattern, i.e., string literal Java, as the first input parameter. The second input parameter, `Pattern.CASE_INSENSITIVE`, enables case-insensitive matching.
+
+Next, we create an object of the `Matcher` class by calling the `Pattern` object's `matcher()` method. We are also passing the text we want to check for matches: if there is a match, the program prints `true`, and `false` if there isn't.
```

Comment:
```suggestion
Next, we create an object of the `Matcher` class by calling the `Pattern` object's `matcher()` method. We are also passing the text we want to check for matches: If there is a match, the program prints `true`, and `false` if there isn't.
```

Reply:
Since there's a colon right before this, are you sure it should be capitalized?

Reply:
Style guides vary on this. Grammarly and other sources say, "When a colon introduces a complete sentence, you may capitalize the first word after the colon according to some style guides. Read on for details..." (https://www.grammarly.com/blog/capitalization-after-colons/) if it looks weird to you, i'm not over-attached to this... feel free to leave as it is.